### PR TITLE
[Form] Don't use field appearances when /NeedAppearances is set to true (bug 1796741)

### PR DESCRIPTION
**Pull request by calixteman**
When a form isn't changed, we used the appearances we had in the file, but when /NeedAppearances is true, all the appearances have to be regenerated whatever they're.

### CODE REVIEW

**Comment by Snuffleupagus** [comment]
I can't access the bug, can you please CC me on it so that I have access to the complete context here?

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js:106]: Given https://github.com/mozilla/pdf.js/blob/987062c302c01bab72645e28a1ffa3fd457e1eb6/src/core/annotation.js#L79 above, could we not simply do something like the following here instead?
```suggestion
        (acroForm && acroForm.get("NeedAppearances)) || false,
```

File: src/core/document.js
Comment (Snuffleupagus) [inline src/core/document.js:644]: With the comment above addressed, do we actually need to make any changes in this file?

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js:1495]: Could we break any other code-path, e.g. related to saving or similar, if we simply clear out the `this.appearance` data here?
Since I don't know the annotation-related code well enough to tell right away, I figured that it couldn't hurt to ask.

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js:151]: Given https://github.com/mozilla/pdf.js/blob/987062c302c01bab72645e28a1ffa3fd457e1eb6/src/core/annotation.js#L79 above, could we not simply do something like the following here instead?
```suggestion
        (acroForm instanceof Dict && acroForm.get("NeedAppearances)) || false,
```

File: src/core/annotation.js
Comment (calixteman) [reply src/core/annotation.js:1488]: Normally no.
When the appearance is null, a new one should be generated from the field value..

**Review by Snuffleupagus — APPROVED** [review]
r=me, with passing tests; thank you!

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js:144]: While there's still a bunch of regressions that we need to address, blocking this PR, I couldn't help to look at these lines and think about the overall readability.

How about we make the following changes, to improve both this and the old code?
```diff
diff --git a/src/core/annotation.js b/src/core/annotation.js
index 748ae7f2b..b7b52e219 100644
--- a/src/core/annotation.js
+++ b/src/core/annotation.js
@@ -119,6 +119,7 @@ class AnnotationFactory {
       return undefined;
     }

+    const acroFormDict = acroForm instanceof Dict ? acroForm : Dict.empty;
     const id =
       ref instanceof Ref ? ref.toString() : `annot_${idFactory.createObjId()}`;

@@ -134,14 +135,12 @@ class AnnotationFactory {
       subtype,
       id,
       pdfManager,
-      acroForm: acroForm instanceof Dict ? acroForm : Dict.empty,
+      acroForm: acroFormDict,
       attachments,
       xfaDatasets,
       collectFields,
       needAppearances:
-        !collectFields &&
-        ((acroForm instanceof Dict && acroForm.get("NeedAppearances")) ||
-          false),
+        !collectFields && acroFormDict.get("NeedAppearances") === true,
       pageIndex,
     };
```

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js:144]: Given that this should be a boolean value, can we please check *explicitly* for that to prevent any future surprises/bugs in corrupt documents?
For example, what if the entry happens to be a non-empty string or any other truthy value? 
```suggestion
        !collectFields && acroFormDict.get("NeedAppearances") === true,
```

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
Revoking approval, while there's still ref-test regressions.

**Comment by calixteman** [comment]
About the "regressions":
 - `issue12504.pdf`: it has the flag `/NeedAppearances` set to true;
 - `issue12750.pdf`: ditto;
 - `annotation-tx.pdf`: ditto;
 - `bug1669099.pdf`: ditto and the newly created borders are slightly different from the ones we had in the stream;
 - `issue12233.pdf`: ditto.

So they aren't regressions at all but only improvements.

**Comment by Snuffleupagus** [comment]
> `annotation-tx.pdf`: ditto;

But the rendering, in the ref-test, is now *double* which makes no sense.
My point here is that something caused the ref-test image to regress, such that it's no longer is a good representation of how the document actually renders/prints in the viewer, and that part needs investigation/fixing here.

**Comment by Snuffleupagus** [comment]
> `annotation-tx.pdf`: ditto;

But the rendering, in the ref-test, is now *double* which makes no sense.
Furthermore you can easily reproduce this in the viewer itself, with the `annotationMode = 1` option set, which means that this is a real regression and not just some test-suite weirdness.
My point here is that something caused the ref-test image to regress, and you such that it's no longer is a good representation of how the document actually renders/prints in the viewer, and that part needs investigation/fixing here.

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js:1711]: Unrelated change, please revert.

File: web/app_options.js
Comment (Snuffleupagus) [inline web/app_options.js:98]: Unrelated change, please revert.

**Comment by calixteman** [comment]
> > `annotation-tx.pdf`: ditto;
> 
> But the rendering, in the ref-test, is now _double_ which makes no sense. Furthermore you can easily reproduce this in the viewer itself, with the `annotationMode = 1` option set, which means that this is a real regression and not just some test-suite weirdness.

It's because of the test itself:
https://github.com/mozilla/pdf.js/blob/098ca670d0f290fbc3f70c66080708a5e4b67155/test/test_manifest.json#L4984

We render the annotation layer:
https://github.com/mozilla/pdf.js/blob/098ca670d0f290fbc3f70c66080708a5e4b67155/test/driver.js#L693

And `annotationMode` is `undefined`:
https://github.com/mozilla/pdf.js/blob/098ca670d0f290fbc3f70c66080708a5e4b67155/test/driver.js#L733

which will cause to have `annotationMode` set to `AnnotationMode.ENABLE`.
So `renderForms` is false:
https://github.com/mozilla/pdf.js/blob/098ca670d0f290fbc3f70c66080708a5e4b67155/src/core/document.js#L462

and finally we don't enter in the block here:
https://github.com/mozilla/pdf.js/blob/098ca670d0f290fbc3f70c66080708a5e4b67155/src/core/annotation.js#L1682

and then the appearance stream is used to render the annotation.

I didn't consider this "regression" as one because for me it was because of the `annotation` type in the test and consequently on purpose.

**Comment by Snuffleupagus** [comment]
> It's because of the test itself:

Sorry, but that still doesn't explain why it can be reproduced *in the viewer* with `annotationMode = 1` set!?

**Comment by Snuffleupagus** [comment]
I still don't at all understand *why* these changes lead to the double rendering here, and I don't think it's correct to just leave this as-is.

(And this PR now looks way to risky to land less than a week before the next release...)

**Comment by calixteman** [comment]
The annotation is rendered one time in the `annotationLayer` and a second time in the canvas because the `intent` parameter in the `getOperatorList` function doesn't contain the flag  `RenderingIntentFlag.ANNOTATIONS_FORMS`.

**Comment by calixteman** [comment]
The pdf has the flag `/NeedAppearances` set to true which means that the appearance for this form field must be generated based on its value, it's why there is something to render.
The stream is null here, so it's why because the patch an empty appearance was "rendered" on the canvas.

**Comment by calixteman** [comment]
Maybe the fix is just to add `RenderingIntentFlag.ANNOTATIONS_FORMS` in the `intent` when we've an annotation layer.
But I think we should do that in a follow-up.

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js:3039]: Won't this cause an error to be thrown below if `annotationStorage === undefined` *and* `this.appearance === undefined`?

Comment (calixteman) [reply src/core/annotation.js]: Yep, you're right, thanks for catching that.
I updated to have something similar to what we have in the appearance getter for text fields.

**Review by Snuffleupagus — APPROVED** [review]
Hopefully we've now fixed all problems here, since the remaining movement in the tests should be expected/improvements.
Given that the double rendering has now been fully understood, and that the fix was pretty simple and straightforward I'm less worried about that.

Your call whether to land this now, or wait until after the upcoming release from the `3`-branch.

**Pull request merged**
